src/database/database-common/src/db_transaction_manager.rs

```diff
@@ -79,17 +75,82 @@ where
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
-#[derive(Debug)]
-pub struct Transaction {
-    pub transaction: Box<dyn Any + Send>,
+/// Represents a shared reference to [`sqlx::Transaction`] that unifies how we
+/// propagate transactions across different database implementations. This
+/// reference can appear in multiple components simultaneously, but can be used
+/// from only one place at a time via locking. Despite its async nature, this
+/// lock must be held only for the duration of the DB query and released when
+/// passing control into other components.
+#[derive(Clone)]
+pub struct TransactionRef {
+    p: Arc<Mutex<Box<dyn Any + Send>>>,
```

Comment:
I'm pretty sure we can get rid of the Box here, as Arc already allocates in heap, but I ran into some issues with downcasting.